placehero/src/components/timeline.rs

```diff
@@ -46,105 +52,123 @@ impl Timeline {
+        // (We plan for the function which makes the future to have access to the app state, but that hasn't happened yet)
+        let user = self.user_id.clone();
+        fork(
+            virtual_scroll(
+                // Show the statuses which have been loaded
+                // We currently never unload previous statuses (note that the widgets but not the remainder are)
+                // +1 to allow room for the spinner.
+                // Note that we also launch the very first initial request from the spinner
+                self.start_index
+                    ..(self.start_index + i64::try_from(self.statuses.len()).unwrap() + 1),
+                |timeline: &mut Self, idx| {
+                    let local_idx = usize::try_from(idx - timeline.start_index).unwrap();
+                    // If we're "close" to the last downloaded item.
+                    if local_idx + BUFFER >= timeline.statuses.len()
+                        // And don't already have a request in flight
+                        && !timeline.pending_id
+                        && !timeline.at_end
+                    {
+                        // Kick off the next request
+                        timeline.pending_id = true;
+                        timeline
+                            .requests
+                            .as_ref()
+                            .unwrap()
+                            .send(TimelineContinuationRequest {
+                                // The max id is the newest status which will be excluded; i.e. we request
+                                // all the statuses from before the last loaded one
+                                // (If we haven't loaded any yet, we just load the first n)
+                                max_id: timeline.statuses.last().map(|it| it.id.clone()),
+                            })
+                            .unwrap();
                     }
-                }
-            },
-            |timeline: &mut Timeline, sender| {
-                timeline.requests = Some(sender);
-                if timeline.statuses.is_empty() {
-                    // TODO: This is a nasty hack
-                    timeline
-                        .requests
-                        .as_ref()
-                        .unwrap()
-                        .send(TimelineContinuationRequest { max_id: None })
-                        .unwrap();
-                }
-            },
-            |timeline: &mut Timeline, resp| {
-                match resp {
-                    Ok(mut instance) => {
-                        // If we're at the oldest post, there's no use making the same request again
-                        if !instance.json.is_empty() {
-                            timeline.pending_id = false;
+                    if local_idx == timeline.statuses.len() {
+                        if timeline.at_end {
+                            OneOf3::A(prose("End of timeline.").text_alignment(TextAlign::Center))
```

Comment:
Why are `OneOf` and `OneOf3` mixed?

Reply:
I believe this to be the idiomatic form (although we've had a few questions about this before), and it's not really documented.

Essentially, we use `OneOf3` as the first variant to let Rust infer that `D to I` are all `Never`, but then use `OneOf` for the remainder. This allows us to add new variants whilst only needing to change the "first" one - not needing to update every item.